### PR TITLE
Improve docs for HorizontalPodAutoscaler

### PR DESCRIPTION
**Pull request by sftim**
(What I hope are) better docs for HPA, following on from PR #30547.

[Preview](https://deploy-preview-30711--kubernetes-io-vnext-staging.netlify.app/docs/tasks/run-application/horizontal-pod-autoscale/), [preview](https://deploy-preview-30711--kubernetes-io-vnext-staging.netlify.app/docs/tasks/run-application/horizontal-pod-autoscale-walkthrough/).

### CODE REVIEW

**Comment by sftim** [comment]
Relevant to issue #4493

File: content/en/docs/tasks/run-application/horizontal-pod-autoscale.md [hunk 1]
```diff
@@ -3,59 +3,77 @@ reviewers:
...
+Horizontal pod autoscaling does not apply to objects that can't be scaled (for example:
+a {{< glossary_tooltip text="DaemonSet" term_id="daemonset" >}}.)
 
+The HorizontalPodAutoscaler is implemented as a Kubernetes API resource and a
+{{< glossary_tooltip text="controller" term_id="controller" >}}.
+The resource determines the behavior of the controller.
+The horizontal pod autoscaling controller, running within the Kubernetes
+{{< glossary_tooltip text="control plane" term_id="control-plane" >}}, periodically adjusts the
+desired scale of its target (for example, a Deployment) to match observed metrics such as average
+CPU utilisation, average memory utilisation,or any other custom metric you specify.
```
Comment (chrisnegus) [inline content/en/docs/tasks/run-application/horizontal-pod-autoscale.md:41]: Is that the British way to spell "utilization"? :-)

Comment (chrisnegus) [reply content/en/docs/tasks/run-application/horizontal-pod-autoscale.md:41]: I think we should use consistent spelling of the term throughout the document, although I don't mind which one you use. Both seem to be correct.

File: content/en/docs/tasks/run-application/horizontal-pod-autoscale.md [hunk 2]
```diff
@@ -89,26 +112,28 @@ desiredReplicas = ceil[currentReplicas * ( currentMetricValue / desiredMetricVal
...
-Before checking the tolerance and deciding on the final values, we take
-pod readiness and missing metrics into consideration, however.
 
-All Pods with a deletion timestamp set (i.e. Pods in the process of being
-shut down) and all failed Pods are discarded.
+Before checking the tolerance and deciding on the final values, the control
+plane also considers whether any metrics are missing, and how many Pods
+are [`Ready`](/docs/concepts/workloads/pods/pod-lifecycle/#pod-conditions).
+All Pods with a deletion timestamp set (objects with a deletion timestamp are
+in in the process of being shut down / removed) are ingored, and all failed Pods
```
Comment (chrisnegus) [inline content/en/docs/tasks/run-application/horizontal-pod-autoscale.md:128]:
```suggestion
in the process of being shut down / removed) are ignored, and all failed Pods
```

File: content/en/docs/tasks/run-application/horizontal-pod-autoscale.md [hunk 3]
```diff
@@ -124,20 +149,21 @@ default is 5 minutes.
...
+Furthermore, if any not-yet-ready pods were present, and the workload would have
+scaled up without factoring in missing metrics or not-yet-ready pods,
+the controller conservatively assumes that the not-yet-ready pods are consuming 0%
+of the desired metric, further dampening the magnitude of a scale up.
 
-After factoring in the not-yet-ready pods and missing metrics, we
-recalculate the usage ratio.  If the new ratio reverses the scale
-direction, or is within the tolerance, we skip scaling.  Otherwise, we use
-the new ratio to scale.
+After factoring in the not-yet-ready pods and missing metrics, the
```
Comment (chrisnegus) [inline content/en/docs/tasks/run-application/horizontal-pod-autoscale.md:162]:
```suggestion
After factoring in the not-yet-ready Pods and missing metrics, the
```

File: content/en/docs/tasks/run-application/horizontal-pod-autoscale.md [hunk 4]
```diff
@@ -296,30 +292,32 @@ Once you have rolled out the container name change to the workload resource, tid
...
-{{< /note >}}
+(the `autoscaling/v2beta2` API version previously provided this ability as a beta feature)
 
-You can also use a HorizontalPodAutoscaler to change the scale of a
-workload based on custom metrics.  You can add custom metrics for the
-Horizontal Pod Autoscaler to use in the `autoscaling/v2` API.
-Kubernetes then queries the new custom metrics API to fetch the values
-of the appropriate custom metrics.
+Provided that you use the `autoscaling/v2` API version, you can configure a HorizontalPodAutoscaler
+to scale based on a custom metric (that is not built in to Kubernetes or any Kubernetes component).
```
Comment (chrisnegus) [inline content/en/docs/tasks/run-application/horizontal-pod-autoscale.md:303]:
```suggestion
to scale based on a custom metric (that is not built into Kubernetes or any Kubernetes component).
```

File: content/en/docs/tasks/run-application/horizontal-pod-autoscale.md [hunk 5]
```diff
@@ -346,20 +343,23 @@ and [external.metrics.k8s.io](https://github.com/kubernetes/community/blob/maste
...
-`scaleDown` section under the `behavior` field. A stabilization window
-can be specified for both directions which prevents the flapping of
-the number of the replicas in the scaling target. Similarly specifying
-scaling policies controls the rate of change of replicas while
-scaling.
+(the `autoscaling/v2beta2` API version previously provided this ability as a beta feature)
 
-### Scaling Policies
+If you use the `v2` HorizontalPodAutoscaler API, you can use the `behavior` field
+(see the [API reference](/docs/reference/kubernetes-api/workload-resources/horizontal-pod-autoscaler-v2/#HorizontalPodAutoscalerSpec)
```
Comment (chrisnegus) [inline content/en/docs/tasks/run-application/horizontal-pod-autoscale.md:353]:
```suggestion
(see the [API reference](/docs/reference/kubernetes-api/workload-resources/horizontal-pod-autoscaler-v2/#HorizontalPodAutoscalerSpec))
```

File: content/en/docs/tasks/run-application/horizontal-pod-autoscale.md [hunk 6]
```diff
@@ -396,21 +396,27 @@ direction. By setting the value to `Min` which would select the policy which all
...
+behavior:
+  scaleDown:
+    stabilizationWindowSeconds: 300
 ```
 
 When the metrics indicate that the target should be scaled down the algorithm looks
-into previously computed desired states and uses the highest value from the specified
-interval. In above example all desired states from the past 5 minutes will be considered.
+into previously computed desired states, and uses the highest value from the specified
+interval. In above example, all desired states from the past 5 minutes will be considered.
```
Comment (chrisnegus) [inline content/en/docs/tasks/run-application/horizontal-pod-autoscale.md:416]:
```suggestion
interval. In the above example, all desired states from the past 5 minutes will be considered.
```

File: content/en/docs/tasks/run-application/horizontal-pod-autoscale.md [hunk 7]
```diff
@@ -509,7 +527,13 @@ replica count or HPA's minimum replica count.
 
 ## {{% heading "whatsnext" %}}
 
+If you configure autoscaling in your cluster, you may also want to consider running a
+cluster-level autoscaler such as [Cluster Autoscaler](https://github.com/kubernetes/autoscaler/tree/master/cluster-autoscaler)
```
Comment (chrisnegus) [inline content/en/docs/tasks/run-application/horizontal-pod-autoscale.md:531]:
```suggestion
cluster-level autoscaler such as [Cluster Autoscaler](https://github.com/kubernetes/autoscaler/tree/master/cluster-autoscaler).
```

File: content/en/docs/tasks/run-application/horizontal-pod-autoscale.md [hunk 8]
```diff
@@ -509,7 +527,13 @@ replica count or HPA's minimum replica count.
...
+For more information on HorizontalPodAutoscaler:
 
-* Design documentation: [Horizontal Pod Autoscaling](https://git.k8s.io/community/contributors/design-proposals/autoscaling/horizontal-pod-autoscaler.md).
-* kubectl autoscale command: [kubectl autoscale](/docs/reference/generated/kubectl/kubectl-commands/#autoscale).
-* Usage example of [Horizontal Pod Autoscaler](/docs/tasks/run-application/horizontal-pod-autoscale-walkthrough/).
+* Read a [walkthrough example](/docs/tasks/run-application/horizontal-pod-autoscale-walkthrough/) for horizontal pod autoscaling.
+* Read documentation for [`kubectl autoscale`](/docs/reference/generated/kubectl/kubectl-commands/#autoscale).
+* If you would like to write your own custom metrics adapter, check out the
+  [boilerplate](https://github.com/kubernetes-sigs/custom-metrics-apiserver) to get started.
+* Read the [API reference](https://kubernetes.io/docs/reference/kubernetes-api/workload-resources/horizontal-pod-autoscaler/) for HorizontalPodAutoscaler
```
Comment (chrisnegus) [inline content/en/docs/tasks/run-application/horizontal-pod-autoscale.md:539]:
```suggestion
* Read the [API reference](https://kubernetes.io/docs/reference/kubernetes-api/workload-resources/horizontal-pod-autoscaler/) for HorizontalPodAutoscaler.
```

**Review by chrisnegus — CHANGES_REQUESTED** [review]
Very nice job cleaning up this document. I like the addition of glossary terms and the use of exact object names. Side note: The some of the official autoscaling reviewers are no longer active. I had trouble finding a reviewer from one of the other PRs I was following. We might want to see about getting that list cleaned up.

File: content/en/docs/tasks/run-application/horizontal-pod-autoscale-walkthrough.md [hunk 9]
```diff
@@ -75,16 +94,27 @@ deployment.apps/php-apache created
...
+[`kubectl autoscale`](/docs/reference/generated/kubectl/kubectl-commands#autoscale) subcommand,
+part of `kubectl`, that helps you do this.
+
+You will shortly run a command that creates a HorizontalPodAutoscaler that maintains
+between 1 and 10 replicas of the Pods controlled by the php-apache Deployment that
+you created in the first step of these instructions.
+
+Roughly speaking, the HPA {{<glossary_tooltip text="controller" term_id="controller">}} will increase and decrease
+the number of replicas (by updating the Deployment) to maintain an average CPU utilization across all Pods of 50%.
+The Deployment then updates the ReplicaSet - this is part of how all Deployments work in Kubernetes -
```
[the remark this answers is not in the document]
Comment (sftim) [reply content/en/docs/tasks/run-application/horizontal-pod-autoscale-walkthrough.md:109]: I meant an en dash with spaces round it (but, I'm not very familiar with American-style punctuation). Could we tweak the punctuation in a follow up PR though?

File: content/en/docs/tasks/run-application/horizontal-pod-autoscale.md (hunk 1 above)
Comment (sftim) [inline content/en/docs/tasks/run-application/horizontal-pod-autoscale.md:41]: We should use US English spellings. Sometimes the English English spellings that I'm accustomed to slip through by mistake.
```suggestion
CPU utilization, average memory utilization, or any other custom metric you specify.
```

File: content/en/docs/tasks/run-application/horizontal-pod-autoscale.md (hunk 4 above)
Comment (sftim) [reply content/en/docs/tasks/run-application/horizontal-pod-autoscale.md:303]: Not sure about this one. The phrase here is “built in”.

File: content/en/docs/tasks/run-application/horizontal-pod-autoscale.md (hunk 6 above)
Comment (sftim) [reply content/en/docs/tasks/run-application/horizontal-pod-autoscale.md:416]: I think this was wrong in the original, FWIW.

File: content/en/docs/tasks/run-application/horizontal-pod-autoscale-walkthrough.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (chrisnegus) [reply content/en/docs/tasks/run-application/horizontal-pod-autoscale-walkthrough.md:109]: As you like. We Americans do things a little different sometimes.

File: content/en/docs/tasks/run-application/horizontal-pod-autoscale.md (hunk 4 above)
Comment (chrisnegus) [reply content/en/docs/tasks/run-application/horizontal-pod-autoscale.md:303]: That's just how it struck me. Fine to leave it.

**Comment by chrisnegus** [comment]
/lgtm
Nice job.

**Comment by jlbutler** [comment]
@gjtempleton could you give this one a quick tech review? thanks!

**Comment by sftim** [comment]
I fixed a spelling mistake I'd accidentally missed.

File: content/en/docs/tasks/run-application/horizontal-pod-autoscale.md [hunk 10]
```diff
@@ -346,20 +343,23 @@ and [external.metrics.k8s.io](https://github.com/kubernetes/community/blob/maste
...
+(the `autoscaling/v2beta2` API version previously provided this ability as a beta feature)
 
-### Scaling Policies
+If you use the `v2` HorizontalPodAutoscaler API, you can use the `behavior` field
+(see the [API reference](/docs/reference/kubernetes-api/workload-resources/horizontal-pod-autoscaler-v2/#HorizontalPodAutoscalerSpec))
+to configure separate scale-up and scale-down behaviors.
+You specify these behaviours by setting `scaleUp` and / or `scaleDown`
+under the `behavior` field.
+
+You can specify a _stabilization window_ that prevents [flapping](#flapping)
```
Comment (gjtempleton) [inline content/en/docs/tasks/run-application/horizontal-pod-autoscale.md:358]: Maybe `mitigates` or `lessens` is better here, as we can't guarantee preventing any scaling flapping.

Comment (gjtempleton) [reply content/en/docs/tasks/run-application/horizontal-pod-autoscale.md:358]: Actually, do we need this segment given largely restating it ~40 lines below?

**Review by gjtempleton — COMMENTED** [review]
Mostly nits, but a number of minor suggestions queries, generally looks like a significant improvement though, thanks!

**Comment by sftim** [comment]
After the v1.23 release I'll see about nudging folk to improve this further.

**Pull request merged**